File: experimental/forthic-cpp/ForthicLib/Words/DefinitionWord.cpp
#include "pch.h"
#include "DefinitionWord.h"
#include "../Interpreter.h"


DefinitionWord::DefinitionWord(string word_name) : Word(word_name)
{
}

DefinitionWord::~DefinitionWord()
{
}

void DefinitionWord::CompileWord(shared_ptr<Word> word)
{
    words.push_back(word);
}

void DefinitionWord::Execute(Interpreter *interp)
{
    for (auto iter = words.begin(); iter != words.end(); iter++)
    {
        (*iter)->Execute(interp);
    }
}

File: forthic-rb/lib/forthic/words/definition_word.rb
# # frozen_string_literal: true

require_relative 'word'
require_relative '../forthic_error'

module Forthic
  class DefinitionWord < Word
    attr_accessor :words, :cur_index

    # @param [String] name
    def initialize(name)
      super(name)
      @words = []
      @cur_index = 0
    end

    # @param [Word] word
    def add_word(word)
      @words.push(word)
    end

    # @param [Interpreter] interp
    def execute(interp)
      @words.each do |word|
        begin
          word.execute(interp)
        rescue => e
          error = ForthicError.new(
            "definition_word-29",
            "Error executing word #{word.name}",
            "Error in #{self.name} definition",
            interp.get_string_location
          )
          error.set_caught_error(e)
          raise error
        end
      end
    end
  end
end